core/templates/dev/head/pages/exploration_editor/editor_tab/StateSolutionHelperService.js

```diff
@@ -0,0 +1,117 @@
+// Copyright 2017 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Service for solution editor.
+ */
+
+oppia.constant('SUPPORTED_INTERACTIONS', [
+  'InteractiveMap',
+  'MusicNotesInput',
+  'GraphInput',
+  'SetInput',
+  'MathExpressionInput',
+  'MultipleChoiceInput',
+  'ImageClickInput',
+  'ItemSelectionInput',
+  'LogicProof'
+]);
+
+oppia.constant('INTERACTION_OBJECT_TYPES', {
+  CodeRepl: 'CodeString',
+  GraphInput: 'Graph',
+  ImageClickInput: 'ImageWithRegions',
+  PencilCodeEditor: 'CodeString',
+  MathExpressionInput: 'UnicodeString',
+  MusicNotesInput: 'MusicPhrase',
+  NumericInput: 'Real',
+  TextInput: 'NormalizedString',
+  LogicProof: 'LogicQuestion'
+});
+
+oppia.factory('StateSolutionHelperService', [
+  'oppiaHtmlEscaper', 'responsesService', 'stateInteractionIdService',
+  'explorationContextService', 'editorContextService', '$injector',
+  'explorationStatesService', 'angularNameService', 'stateSolutionService',
+  'AnswerClassificationService', 'alertsService',
+  'SUPPORTED_INTERACTIONS', 'INTERACTION_OBJECT_TYPES',
+  function(oppiaHtmlEscaper, responsesService, stateInteractionIdService,
+    explorationStatesService, editorContextService, $injector,
+    explorationContextService, angularNameService, stateSolutionService,
+    AnswerClassificationService, alertsService,
+    SUPPORTED_INTERACTIONS, INTERACTION_OBJECT_TYPES) {
+    var currentSolutionIsValid = true;
```

Comment:
Services are singletons and don't refer to any particular state -- so what "current solution" does this refer to? You're probably going to end up mixing data between states if you do it like this...unless you're very very careful about bookkeeping, but this still seems like a maintainability issue to me.

If you need to keep track of validity, the interaction object is probably the best place to do it, since it is the lowest common ancestor of hints and solutions. That would also solve the warnings problem since you can just update that flag on any update to answer groups / hints, and during warning computation you could just read off the variable there.